core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -74,255 +61,286 @@
    }
 
    @Override
    @Stop
    public void stop() {
-      storesMutex.writeLock().lock();
-      publisherSemaphore.acquireUninterruptibly(Integer.MAX_VALUE);
+      long stamp = lock.writeLock();
       try {
+         stopAvailabilityTask();
+         Flowable<NonBlockingStore<?, ?>> flowable = Flowable.fromIterable(stores.keySet());
          // If needed, clear the persistent store before stopping
          if (clearOnStop) {
-            clearAllStoresSync(AccessMode.BOTH, getNextTraceNumber());
+            flowable = flowable
+                  .delay(store -> Completable.fromCompletionStage(store.clear()).toFlowable());
          }
+         flowable = flowable.delay(store -> Completable.fromCompletionStage(store.stop()).toFlowable());
 
-         Set<Lifecycle> undelegated = new HashSet<>();
-         Consumer<CacheWriter> stopWriters = writer -> {
-            writer.stop();
-            if (writer instanceof DelegatingCacheWriter) {
-               CacheWriter actual = undelegate(writer);
-               actual.stop();
-               undelegated.add(actual);
-            } else {
-               undelegated.add(writer);
-            }
-         };
-         if (availabilityFuture != null)
-            availabilityFuture.cancel(true);
-         nonTxWriters.forEach(stopWriters);
-         nonTxWriters.clear();
-         txWriters.forEach(stopWriters);
-         txWriters.clear();
-
-         for (CacheLoader l : loaders) {
-            if (!undelegated.contains(l)) {
-               l.stop();
-            }
-            if (l instanceof DelegatingCacheLoader) {
-               CacheLoader actual = undelegate(l);
-               if (!undelegated.contains(actual)) {
-                  actual.stop();
-               }
-            }
-         }
-         loaders.clear();
+         // Wait until it completes
+         blockingSubscribe(flowable);
+         stores.clear();
          preloaded = false;
       } finally {
-         publisherSemaphore.release(Integer.MAX_VALUE);
-         storesMutex.writeLock().unlock();
+         lock.unlockWrite(stamp);
       }
    }
 
-   @Override
-   public boolean hasWriter() {
-      if (!enabled) {
-         return false;
-      }
-      acquireReadLock();
-      try {
-         return !nonTxWriters.isEmpty() || !txWriters.isEmpty();
-      } finally {
-         releaseReadLock();
+   private void stopAvailabilityTask() {
+      AutoCloseable taskToClose = availabilityTask;
+      if (taskToClose != null) {
+         try {
+            taskToClose.close();
+         } catch (Exception e) {
+            log.warn("There was a problem stopping availability task", e);
+         }
       }
    }
 
+   // This here solely to document that we are using a blocking method. This is because the start/stop lifecycle
+   // methods themselves are blocking but our API is not. This can be removed if lifecycle ever allows for non
+   // blocking, but don't hold your breath for it.
```

Comment:
Do we have a Jira for this? \cc @danberindei

Reply:
Not that I am aware of. I am guessing we may have this in Infinispan 14 🤣

Reply:
That sounds optimistic :smiley: